Added lifecycle application module

Sgitario opened the pull request:
Added basic coverage for https://issues.redhat.com/browse/QUARKUS-866.

Note that this module is not working due to the above issue.
I still need to either disable some tests or add the workaround.

rsvoboda commented on lifecycle-application/src/test/java/io/quarkus/ts/openshift/lifecycle/LifecycleApplicationOpenShiftIT.java:
```diff
@@ -0,0 +1,42 @@
+@OpenShiftTest
+public class LifecycleApplicationOpenShiftIT {
+
+    private static final String DEFAULT_ARGUMENT = "-Djava.util.logging.manager=org.jboss.logmanager.LogManager";
```
this looks suspicious, is `-Djava.util.logging.manager=org.jboss.logmanager.LogManager` added to java command before `-jar ....` or after?

Sgitario replied:
I'm not sure what's your point here. Here, we're verifying that `@CommandLineArguments` is injecting the argument properly and these arguments are provided by the OpenShift extension.

rsvoboda replied:
What's the command executed on OpenShift?

Is it `java -Djava.util.logging.manager=org.jboss.logmanager.LogManager -jar quarkus.jar` or is it `java -jar quarkus.jar -Djava.util.logging.manager=org.jboss.logmanager.LogManager` ?

Sgitario replied:
Oh got it. Thanks for spotting it. 
I've changed this test case to actually assert that the logging property is not an argument.

rsvoboda commented on lifecycle-application/src/test/java/io/quarkus/ts/openshift/lifecycle/LifecycleApplicationOpenShiftIT.java:
```diff
@@ -0,0 +1,42 @@
+
+        String actualLog = openshiftClient.pods().withName(pods.getItems().get(0).getMetadata().getName()).getLog();
+
+        assertTrue(actualLog.contains(Main.MESSAGE), "Pod log does not contain the expected message from Main.java");
```
there should be also check for `Received arguments: ...` to ensure nothing gets sneaked in.

And if the tests fails it would be good to see the content of `actualLog`

Sgitario replied:
Changed.

rsvoboda commented on lifecycle-application/src/test/java/io/quarkus/ts/openshift/lifecycle/LifecycleApplicationOpenShiftIT.java:
```diff
@@ -0,0 +1,48 @@
+
+        String actualLog = openshiftClient.pods().withName(pods.getItems().get(0).getMetadata().getName()).getLog();
+
+        Optional<String> argumentsLine = actualLog.lines().filter(line -> line.contains("Received arguments:")).findFirst();
```
`actualLog.lines()` is JDK 11 API and the TS is set to 1.8.

We run it with Java 11, not so sure if we should switch to 11 just because this one line.
If there are more places that could benefit from Java 11 APIs, we should do the switch.

rsvoboda replied:
`"Received arguments:"` could be a constant defined as private static final String

Sgitario replied:
Changed

rsvoboda commented on lifecycle-application/src/test/java/io/quarkus/ts/openshift/lifecycle/LifecycleApplicationOpenShiftIT.java:
```diff
@@ -0,0 +1,48 @@
+        Optional<String> argumentsLine = actualLog.lines().filter(line -> line.contains("Received arguments:")).findFirst();
+        assertTrue(argumentsLine.isPresent(), "Pod log does not contain the received arguments. Actual content: " + actualLog);
+        assertFalse(argumentsLine.get().contains(LOGGING_PROPERTY),
+                "Pod log contain wrong arguments. Actual content: " + actualLog);
```
`argumentsLine` should be enough, instead of `actualLog`

Sgitario replied:
Good idea!

rsvoboda commented on lifecycle-application/src/test/java/io/quarkus/ts/openshift/lifecycle/LifecycleApplicationOpenShiftIT.java:
```diff
@@ -0,0 +1,48 @@
+        .then().statusCode(200)
+                .body(not(containsString(LOGGING_PROPERTY)));
+    }
+
```
Should we have tests that ensures certain format of the executed command?
That test would be dependent on the use mode - JVM or Native.
We would probably check bind address definition and logging property.

Well, how does this app behave in native mode? How are args differentiated from quarkus `-D` settings?

rsvoboda replied:
So I checked and `target/lifecycle-application-1.0.0-SNAPSHOT-runner -Daaa=bbb foo bar` translates to `Received arguments: foo,bar`.

So anything in `-Dfoo=bar` format get transformed as property and is not passed further as argument for the application.

rsvoboda replied:
ATM there is no direct demand to ensure certain format of the executed command. I think we can leave it as possible enhancement for the future.

Sgitario replied:
This makes sense but it's now a limitation when deploying into OpenShift as we can't provide arguments now :/

Sgitario commented:
@rsvoboda , I think I've already addressed your comments. Also, I've created a new assertion to validate expected arguments. Note that I've commented this verification out as I could not find any property to do this.

Sgitario commented:
Added "DO NOT MERGE" label to state that this is not working for upstream issues that will be addressed soon.

Sgitario commented:
run tests

Sgitario commented:
run tests

Sgitario commented:
run tests

Sgitario commented:
run tests

Sgitario commented:
run tests

Sgitario commented:
QUARKUS-866 has been fixed in upstream, but this PR remains blocked because of https://github.com/quarkusio/quarkus/issues/15953

rsvoboda commented:
pls rebase

Sgitario commented:
> pls rebase

Done

Sgitario commented:
Depends on https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/266

Sgitario commented:
Rebased and verified that is correctly working on OCP. 
This PR can be reviewed/merged as dependant PR has been merged too.
